docs/explanation/architectures.rst

```diff
@@ -0,0 +1,168 @@
+* ``core20`` uses ``build-on`` and ``run-on``
+
+build-on
+^^^^^^^^
+
+The architecture of the machine that the snap is built on (i.e. the machine
+where snapcraft runs).
+
+build-for
+^^^^^^^^^
+
+The architecture of the machine that the snap is built for (i.e. the machine
+where the snap will be installed).
+
+run-on
+^^^^^^
+
+This is used for ``core20`` and has the same meaning as ``build-for``. It has
+been replaced with the preferred term ``build-for`` in ``core22``.
+
+.. _build-plans:
+
+Build plans
+-----------
+
+A build plan is a list of what snaps snapcraft will build. It can be defined
+in the ``snapcraft.yaml`` and filtered with command-line arguments and
+environment variables.
+
+core22
+^^^^^^
+
+Snapcraft can create multiple snaps, each built for a different architecture.
+Consider the following ``snapcraft.yaml`` snippet:
+
+.. code-block:: yaml
+
+  architectures:
+    - build-on: amd64
+      build-for: amd64
+    - build-on: [amd64, arm64]
+      build-for: arm64
+
+If snapcraft executes on an ``amd64`` machine, then it will create the
+following build plan:
+
+.. code-block:: text
+
+  Created build plan:
+    build-on: amd64 build-for: amd64
+    build-on: amd64 build-for: arm64
+
+Two snap files will be created: ``my-snap_1.0_amd64.snap`` and
+``my-snap_1.0_arm64.snap``.
+
+This build plan can be filtered with the environment variable
+``SNAPCRAFT_BUILD_FOR=<arch>`` or the command-line argument
+``--build-for=<arch>``. The command-line argument takes priority over the
+environment variable. In the example above, using ``--build-for=arm64`` would
+cause snapcraft to only build one snap for ``arm64``.
+
+Building with a provider
+""""""""""""""""""""""""
+
+When building a snap with LXD or Multipass, each build in the build plan occurs
+in its own environment.
+
+Destructive mode
+""""""""""""""""
+
+In destructive mode, all builds in the build plan occur in the same location.
+This can cause unintended consequences, such as parts not being re-built. For
+more information, see `this issue`_.
+
+To workaround this, use ``--build-for`` or ``SNAPCRAFT_BUILD_FOR`` to build
+one snap at a time and run ``snapcraft clean --destructive-mode`` when changing
+the build-for architecture.
+
+core20
+^^^^^^
+
+Build plans are not supported in ``core20`` so building a ``core20`` snap will
+only produce one snap.
+
+Snapcraft does not automatically clean the build environment when the
+``architecture`` keyword is changed. Therefore ``snapcraft clean`` should be
+run when changing architectures.
+
+Remote build
+^^^^^^^^^^^^
+
+Launchpad supports building snaps on multiple architectures.
+
+If architectures are not defined in the ``snapcraft.yaml``, then Launchpad will
+build the snap on all architectures
+:ref:`supported by Launchpad<supported-architectures-launchpad>`.
+
+If architectures are defined in the ``snapcraft.yaml``, then Launchpad will
+build the snap on all ``build-on`` architectures.
```

Comment:
If I specify:

```yaml
  architectures:
    - build-on: [amd64, arm64]
      build-for: arm64
```

how will Launchpad behave?

I presume the following, but I haven't checked:

- Only one build for each `build-for` architecture
- Prefers native builds over cross-compiles.

What if I say:

```yaml
architectures:
  - build-on: [amd64, arm64]
    build-for: ppc64el
```

Is there a preferred order for `build-on` platforms? (Or maybe it runs on the next available builder for any of those platforms?)

(If it's unknown/undefined I don't mind leaving it unsaid, but if we have a definitive answer it might be useful to include.)

Reply:
Great question!  Launchpad has it's own logic for determining bases.  It's somewhere in here: [[1](https://git.launchpad.net/launchpad/tree/lib/lp/snappy/model/snap.py#n874)] [[2](https://git.launchpad.net/launchpad/tree/lib/lp/snappy/model/snap.py#n938)] [[3](https://git.launchpad.net/launchpad/tree/lib/lp/snappy/adapters/buildarch.py)].  I'm going to dig through that code later.

In the meantime, I've made it clear in the docs that this is ambiguous and outside the user's control.